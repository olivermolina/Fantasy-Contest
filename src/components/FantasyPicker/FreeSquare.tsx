import React, { useMemo, useState } from 'react';
import dayjs from 'dayjs';
import duration from 'dayjs/plugin/duration';
import utc from 'dayjs/plugin/utc';
import timezone from 'dayjs/plugin/timezone';
import { Skeleton } from '@mui/material';

dayjs.extend(utc);
dayjs.extend(timezone);
dayjs.extend(duration);

export interface FreeSquareProps {
  /**
   * Free square discount
   * @example 95
   */
  discount: number;
  /**
   * Game date time as the expiry of the free square promo
   * @example '2022-10-02 13:35:01',
   */
  gameDateTime: string;
}

/**
 * The Free Square Box promotion on top of the FantasyCard component
 */
export default function FreeSquare(props: FreeSquareProps) {
  const { gameDateTime, discount } = props;

  const [time, setTime] = useState<string>();

  useMemo(() => {
    const currentTime = dayjs();
    const gameTimeTz = dayjs(gameDateTime).tz('America/New_York', true);
    const diffTime = dayjs(gameTimeTz).unix() - currentTime.unix();

    let duration = dayjs.duration(diffTime * 1000, 'milliseconds');
    const interval = 1000;
    const twoDP = (n: number) => (n > 9 ? n : '0' + n);

    setInterval(function () {
      duration = dayjs.duration(
        duration.asMilliseconds() - interval,
        'milliseconds',
      );

      const days = duration.days() > 0 ? `${twoDP(duration.days()) + 'd'}` : '';
      const hours = duration.hours() + 'h';
      const minutes = twoDP(duration.minutes()) + 'm';
      const seconds = twoDP(duration.seconds()) + 's';

      const timestamp = `${days} ${hours} ${minutes} ${seconds}`;
      setTime(timestamp);
    }, interval);
  }, [gameDateTime]);

  return (
    <div
      className={
        'flex justify-evenly p-1 rounded-t-lg bg-yellow-600 font-bold text-white items-center text-xs sm:text-sm md:text-md lg:text-lg'
      }
    >
      {time ? (
        <span
          className={
            'bg-black text-white px-1 lg:px-2 rounded-sm whitespace-nowrap'
          }
        >
          {time}
        </span>
      ) : (
        <Skeleton sx={{ width: 150 }} />
      )}
      <p className={'whitespace-nowrap text-black'}>{discount}% OFF</p>
    </div>
  );
}
