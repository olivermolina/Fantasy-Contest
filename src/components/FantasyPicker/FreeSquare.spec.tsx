import React from 'react';
import { render } from '@testing-library/react';
import FreeSquare from './FreeSquare';

describe('FreeSquare', () => {
  it('should render correctly', () => {
    const props = {
      gameDateTime: '2023-03-25 8:00 PM',
      discount: 95,
    };
    const { container } = render(<FreeSquare {...props} />);
    expect(container).toMatchSnapshot();
  });
});
