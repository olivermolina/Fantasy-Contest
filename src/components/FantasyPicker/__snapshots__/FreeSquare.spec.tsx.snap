// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FreeSquare should render correctly 1`] = `
<div>
  <div
    class="flex justify-evenly p-1 rounded-t-lg bg-yellow-600 font-bold text-md lg:text-md text-white "
  >
    <span
      class="MuiSkeleton-root MuiSkeleton-text MuiSkeleton-pulse css-1staz2m-MuiSkeleton-root"
    />
    <span
      class="whitespace-nowrap text-black"
    >
      95
      % OFF
    </span>
  </div>
</div>
`;
